plugin/plugins/s3ds/s3ds.go

```diff
@@ -0,0 +1,175 @@
+package s3ds
+
+import (
+	"fmt"
+	"os"
+
+	s3 "gx/ipfs/QmR1J2FeuQqVZwigFA8e1NPXWacgKGZ3uV8ikqzHua6Pm3/s3gof3r"
+	ds "gx/ipfs/QmSpg1CvpXQQow5ernt1gNBXaXV6yxyNqi7XoeerWfzB5w/go-datastore"
+	"gx/ipfs/QmSpg1CvpXQQow5ernt1gNBXaXV6yxyNqi7XoeerWfzB5w/go-datastore/query"
+	s3ds "gx/ipfs/QmUXSbVz5SggPB2SAVGvzVwo2SWVPqx5oYgLbx2EiTfnsL/go-ds-s3"
+
+	"github.com/ipfs/go-ipfs/plugin"
+	"github.com/ipfs/go-ipfs/repo"
+	"github.com/ipfs/go-ipfs/repo/fsrepo"
+)
+
+var Plugins = []plugin.Plugin{
+	&s3Plugin{},
+}
+
+type s3Plugin struct{}
+
+var _ plugin.PluginDatastore = (*s3Plugin)(nil)
+
+func (s3p s3Plugin) Name() string {
+	return "ds-s3"
+}
+
+func (s3p s3Plugin) Version() string {
+	return "0.0.1"
+}
+
+func (s3p s3Plugin) Init() error {
+	return nil
+}
+
+func (s3p s3Plugin) DatastoreTypeName() string {
+	return "s3ds"
+}
+
+func (s3p s3Plugin) DatastoreConfigParser() fsrepo.ConfigFromMap {
+	return func(m map[string]interface{}) (fsrepo.DatastoreConfig, error) {
+		domain, ok := m["domain"].(string)
+		if !ok {
+			return nil, fmt.Errorf("s3ds: no domain specified")
+		}
+
+		bucket, ok := m["bucket"].(string)
+		if !ok {
+			return nil, fmt.Errorf("s3ds: no bucket specified")
+		}
+
+		rootDir, ok := m["rootDir"].(string)
+		if !ok {
+			return nil, fmt.Errorf("s3ds: no rootDir specified")
+		}
+
+		keys, err := s3.InstanceKeys()
+		if err != nil {
+			keys, _ = s3.EnvKeys()
+
+			// Standardized env variables use AWS_SESSION_TOKEN instead of AWS_SECURITY_TOKEN
+			// See: https://aws.amazon.com/blogs/security/a-new-and-standardized-way-to-manage-credentials-in-the-aws-sdks/
+			//
+			// Should also move from:
+			// https://github.com/rlmcpherson/s3gof3r
+			// to the official aws-sdk-go:
+			// https://github.com/aws/aws-sdk-go-v2
+			sessionToken := os.Getenv("AWS_SESSION_TOKEN")
+			if sessionToken != "" {
+				keys.SecurityToken = sessionToken
+			}
+		}
+
+		accessKey, ok := m["accessKey"].(string)
+		if ok {
+			keys.AccessKey = accessKey
+		}
+
+		secretKey, ok := m["secretKey"].(string)
+		if ok {
+			keys.SecretKey = secretKey
+		}
+
+		sessionToken, ok := m["sessionToken"].(string)
+		if ok {
+			keys.SecurityToken = sessionToken
+		}
+
+		return &S3Config{
+			cfg: &s3ds.Config{
+				Domain:       domain,
+				Bucket:       bucket,
+				RootDir:      rootDir,
+				AccessKey:    keys.AccessKey,
+				SecretKey:    keys.SecretKey,
+				SessionToken: keys.SecurityToken,
+			},
+		}, nil
+	}
+}
+
+type S3Config struct {
+	cfg *s3ds.Config
+}
+
+func (s3c *S3Config) DiskSpec() fsrepo.DiskSpec {
+	return map[string]interface{}{
+		"domain": s3c.cfg.Domain,
+		"bucket": s3c.cfg.Bucket,
+	}
+}
+
+func (s3c *S3Config) Create(path string) (repo.Datastore, error) {
+	return &loggingDatastore{
+		Datastore: s3ds.NewS3Datastore(s3c.cfg),
+	}, nil
+}
+
+type loggingDatastore struct {
```

Comment:
There already is a logging datastore: https://github.com/ipfs/go-ipfs/blob/master/repo/fsrepo/datastores.go#L66

Reply:
Thanks, just starting to learn the codebase.